### DISCUSSION TITLE
Custom hovertool extension: this._slots does not exist

### DISCUSSION THREAD

**Discussion by zcyounger** [bokeh/bokeh#14001] [General]
I want to create my own hovertool extension. To do this, I first copy the Typescript code in [bokeh/bokehjs/src/lib/models/tools/inspectors/hover_tool.ts at 3.1.1 · bokeh/bokeh · GitHub 1](https://github.com/bokeh/bokeh/blob/3.1.1/bokehjs/src/lib/models/tools/inspectors/hover_tool.ts) and do some change:

1. relative import path → absolute import path
2. HoverToolView → myHoverToolView
3. HoverTool → myHoverTool
4. this.register_alias(“hover”, () => new HoverTool()) ->this.register_alias(“myhover”, () => new myHoverTool())
5. override tool_name = “Hover” → override tool_name = “myHover”

The code can be seen here!
https://discourse.bokeh.org/t/custom-hovertool-extension-this-slots-does-not-exist/11723

I get a compilation failed:
myHoverTool.ts:178:23 - error TS2339: Property ‘_slots’ does not exist on type ‘myHoverToolView’.
178 const slot = this._slots.get(this.update)

However, I confirmed that Property ‘_slots’ exists on its basis class View

**Comment by bryevdv** [comment]
@zcyounger please edit the post to fix the code formatting. 

cc @mattpap for input

**Comment by mattpap** [comment]
`View._slots` is an `@internal` JS property, which means it exists at runtime but it's removed from generated declaration files, thus the error. Looking at the code on discourse, you've hit the only case where `_slots` is being used in derived classes. I suppose that's why it's declared as a protected internal and not a private property. For now you can just change the code to:
```ts
const slot = (this as any)._slots.get(this.update)
```
or add 
```ts
declare _slots: WeakMap<Slot<any, any>, Slot<any, any>>
```
at class member level, to avoid the error.

We need to modernize this code, so that if you copy the code, you don't have to deal with such low-level details of `View`'s implementation.

**Reply by zcyounger** [reply]
Thank you! I have solved my problem in your way!
